db/dcrpg/internal/stakestmts.go

```diff
@@ -0,0 +1,218 @@
+		WHERE id IN (SELECT id FROM (
+				SELECT id, ROW_NUMBER()
+				OVER (partition BY tx_hash, block_hash ORDER BY id) AS rnum
+				FROM tickets) t
+			WHERE t.rnum > 1);`
+
+	// Votes
+
+	CreateVotesTable = `CREATE TABLE IF NOT EXISTS votes (
+		id SERIAL PRIMARY KEY,
+		height INT4,
+		tx_hash TEXT NOT NULL,
+		block_hash TEXT NOT NULL,
+		candidate_block_hash TEXT NOT NULL,
+		version INT2,
+		vote_bits INT2,
+		block_valid BOOLEAN,
+		ticket_hash TEXT,
+		ticket_tx_db_id INT8,
+		ticket_price FLOAT8,
+		vote_reward FLOAT8
+	);`
+
+	// Insert
+	insertVoteRow0 = `INSERT INTO votes (
+		height, tx_hash,
+		block_hash, candidate_block_hash,
+		version, vote_bits, block_valid,
+		ticket_hash, ticket_tx_db_id, ticket_price, vote_reward)
+	VALUES (
+		$1, $2,
+		$3, $4,
+		$5, $6, $7,
+		$8, $9, $10, $11) `
+	insertVoteRow = insertVoteRow0 + `RETURNING id;`
+	// insertVoteRowChecked = insertVoteRow0 + `ON CONFLICT (tx_hash, block_hash) DO NOTHING RETURNING id;`
+	upsertVoteRow = insertVoteRow0 + `ON CONFLICT (tx_hash, block_hash) DO UPDATE 
+		SET tx_hash = $3, block_hash = $4 RETURNING id;`
+	insertVoteRowReturnId = `WITH ins AS (` +
+		insertVoteRow0 +
+		`ON CONFLICT (tx_hash, block_hash) DO UPDATE
+		SET tx_hash = NULL WHERE FALSE
+		RETURNING id
+		)
+	SELECT id FROM ins
+	UNION  ALL
+	SELECT id FROM votes
+	WHERE  tx_hash = $3 AND block_hash = $4
+	LIMIT  1;`
+
+	SelectAllVoteDbIDsHeightsTicketHashes = `SELECT id, height, ticket_hash FROM votes;`
+	SelectAllVoteDbIDsHeightsTicketDbIDs  = `SELECT id, height, ticket_tx_db_id FROM votes;`
+
+	// Index
+	IndexVotesTableOnHashes = `CREATE UNIQUE INDEX uix_votes_hashes_index
+		ON votes(tx_hash, block_hash);`
+	DeindexVotesTableOnHashes = `DROP INDEX uix_votes_hashes_index;`
+
+	IndexVotesTableOnCandidate = `CREATE INDEX uix_votes_candidate_block
+		ON votes(candidate_block_hash);`
+	DeindexVotesTableOnCandidate = `DROP INDEX uix_votes_candidate_block;`
+
+	IndexVotesTableOnVoteVersion = `CREATE INDEX uix_votes_vote_version
+		ON votes(version);`
+	DeindexVotesTableOnVoteVersion = `DROP INDEX uix_votes_vote_version;`
+
+	DeleteVotesDuplicateRows = `DELETE FROM votes
+		WHERE id IN (SELECT id FROM (
+				SELECT id, ROW_NUMBER()
+				OVER (partition BY tx_hash, block_hash ORDER BY id) AS rnum
+				FROM votes) t
+			WHERE t.rnum > 1);`
+
+	// Misses
+
+	CreateMissesTable = `CREATE TABLE IF NOT EXISTS misses (
+		id SERIAL PRIMARY KEY,
+		height INT4,
+		block_hash TEXT NOT NULL,
+		candidate_block_hash TEXT NOT NULL,
+		ticket_hash TEXT NOT NULL
+	);`
+
+	// Insert
+	insertMissRow0 = `INSERT INTO misses (
+		height, block_hash, candidate_block_hash, ticket_hash)
+	VALUES (
+		$1, $2, $3, $4) `
+	insertMissRow = insertMissRow0 + `RETURNING id;`
+	// insertVoteRowChecked = insertMissRow0 + `ON CONFLICT (ticket_hash, block_hash) DO NOTHING RETURNING id;`
+	upsertMissRow = insertMissRow0 + `ON CONFLICT (ticket_hash, block_hash) DO UPDATE 
+		SET ticket_hash = $4, block_hash = $2 RETURNING id;`
```

Comment:
should really be a do nothing, but the `...DO NOTHING RETURNING Id;` bit from the commented line wasn't working properly